vortex-layout/src/layouts/row_id.rs

```diff
@@ -0,0 +1,366 @@
+use vortex_dtype::{DType, FieldMask};
+use vortex_error::{VortexExpect, VortexResult};
+use vortex_expr::transform::var_partition::{VarPartitionedExpr, var_partitions_with_map};
+use vortex_expr::{ExactExpr, ExprRef, IDENTITY_IDENTIFIER, Identifier, Scope, ScopeDType};
+use vortex_mask::Mask;
+use vortex_sequence::SequenceArray;
+
+use crate::{ArrayEvaluation, LayoutReader, LayoutReaderRef, MaskEvaluation, PruningEvaluation};
+
+pub struct RowIdLayoutReader {
+    child: LayoutReaderRef,
+    name: Arc<str>,
+    partitioned_expr_cache: DashMap<ExactExpr, Arc<VarPartitionedExpr>>,
+}
+
+static ROW_ID: LazyLock<Identifier> = LazyLock::new(|| Arc::from("row_id"));
+
+impl RowIdLayoutReader {
+    pub fn new(child: LayoutReaderRef) -> Self {
+        Self {
+            child,
+            name: Arc::from("row_id_layout_reader"),
+            partitioned_expr_cache: Default::default(),
+        }
+    }
+}
+
+impl RowIdLayoutReader {
+    /// Utility for partitioning an expression over the fields of a struct.
+    fn partition_expr(&self, expr: &ExprRef) -> Arc<VarPartitionedExpr> {
+        self.partitioned_expr_cache
+            .entry(ExactExpr(expr.clone()))
+            .or_insert_with(|| {
+                // Partition the expression into expressions that can be evaluated over the row_id field
+                // and all other fields that a delegated to their children.
+                Arc::new(
+                    var_partitions_with_map(expr, |id| {
+                        if *id == *ROW_ID {
+                            ROW_ID.clone()
+                        } else {
+                            "".into()
+                        }
+                    })
+                    .vortex_expect("We should not fail to partition variables"),
+                )
+            })
+            .clone()
+    }
+}
+
+impl LayoutReader for RowIdLayoutReader {
+    fn name(&self) -> &Arc<str> {
+        &self.name
+    }
+
+    fn dtype(&self) -> &DType {
+        self.child.dtype()
+    }
+
+    fn row_count(&self) -> Precision<u64> {
+        self.child.row_count()
+    }
+
+    fn register_splits(
+        &self,
+        field_mask: &[FieldMask],
+        row_offset: u64,
+        splits: &mut BTreeSet<u64>,
+    ) -> VortexResult<()> {
+        self.child.register_splits(field_mask, row_offset, splits)
+    }
+
+    fn pruning_evaluation(
+        &self,
+        row_range: &Range<u64>,
+        expr: &ExprRef,
+    ) -> VortexResult<Box<dyn PruningEvaluation>> {
+        // TODO(joe): add variable and then row_id to the pruning eval
+        self.child.pruning_evaluation(row_range, expr)
+    }
+
+    fn filter_evaluation(
+        &self,
+        row_range: &Range<u64>,
+        expr: &ExprRef,
+    ) -> VortexResult<Box<dyn MaskEvaluation>> {
+        let arr_len = row_range.clone().count();
+        let partitioned = self.partition_expr(expr);
+        let Some(row_id) = partitioned.find_partition(&ROW_ID) else {
```

Comment:
If the array already contains a `row_id` column I suspect we will override it with the sequence array in our evaluations?